server/observer.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/grpc"
+)
+
+// UnaryObserver provides an interface for observing unary requests and responses.
+type UnaryObserver interface {
```

Comment:
I think I know what you're trying to do here, but I think it is confusingly similar to UnaryServerInterceptor.

My current opinion, would it be better (or more flexible) to create a helper like this?
```
func UnaryServerChain(... grpc.UnaryServerInterceptor) grpc.UnaryServerInterceptor
```
Then it's up to the users to work on things themselves. Instead of adding/deleting interceptors, they can arrange for interceptors to enable/disable themselves as well.

Reply:
Agreed but chaining `UnaryServerInterceptors` is not possible.

It is explicitely stated in the grpc documentation: https://godoc.org/google.golang.org/grpc#UnaryInterceptor.

The reason why chaining is not possible is because the handler passed in to the interceptor should only be called once since the handler completes the normal execution of a unary RPC.

Reply:
I think you misunderstood me. I meant, we can wrap a list of user-provided interceptors, do the chaining logic ourselves, like you did it here. Then present grpc with only on interceptor. Am I making sense? I believe that should work right?

Reply:
Yeah I think that strategy could work. We'd have to mock the handler to ensure that the actual handler only gets called once.

```go

func UnaryServerChain(interceptors ...grpc.UnaryServerInterceptor) grpc.UnaryServerInterceptor {
	return func(ctx context.Context, req interface{}, info *UnaryServerInfo, handler UnaryHandler) (interface{}, error) {
		resp, err := handler(ctx, req)
		mockHandler := func(context.Context, interface{}) (interface{}, error) { return resp, err }
		for _, interceptor := range interceptors {
			interceptor(ctx, req, info, mockHandler)
		}
		return resp, err
	}
}
```

For UnaryInterceptors I think this makes sense. But to do something similar for stream interceptors will cause a lot of extra code down the line. Every test case will need to implement a `grpc.ServerStream` to hook into the streaming request and responses. 

I'd argue to keep doing the stream observations this way, to keep that interface simple and then to keep the observers this way to have them be consistent.

Reply:
I think that sounds reasonable :P

Reply:
Woops I had a typo, 

Keep stream observations this way and keep the unary observations this way to keep them consistent.